### ISSUE TITLE
0.4: TypeError when requesting a list in the admin UI (/keystone/pages)

### ISSUE THREAD

**Issue by yn5** [keystonejs/keystone#2081]
I get the following Type Error requesting a list in the admin UI (/keystone/pages) on current master: 53b710a321cdcc1978e2ec93e8ddce7e1055bcb4

TypeError: /Users/yn5/stuff/web/puha/keystone/lib/core/../../admin/server/templates/list.jade:30
    28| 			Keystone.showCreateForm = !{serializeJavaScript(showCreateForm)};
    29| 			Keystone.createFormData = !{serializeJavaScript(submitted)};
  > 30| 			Keystone.createFormErrors = !{serializeJavaScript(createErrors || null)};
    31| 		script(src='#{adminPath}/js/packages.js')
    32| 		script(src='#{adminPath}/js/fields.js')
    33| 		script(src='#{adminPath}/js/list.js')

Serializing native function: now
    at /Users/yn5/stuff/web/puha/keystone/node_modules/serialize-javascript/index.js:77:19
    at String.replace (native)
    at serialize (/Users/yn5/stuff/web/puha/keystone/node_modules/serialize-javascript/index.js:68:16)
    at eval (eval at <anonymous> (/Users/yn5/stuff/web/puha/keystone/node_modules/jade/lib/index.js:218:8), <anonymous>:529:49)
    at eval (eval at <anonymous> (/Users/yn5/stuff/web/puha/keystone/node_modules/jade/lib/index.js:218:8), <anonymous>:621:22)
    at res (/Users/yn5/stuff/web/puha/keystone/node_modules/jade/lib/index.js:219:38)
    at render (/Users/yn5/stuff/web/puha/keystone/lib/core/render.js:111:11)
    at renderView (/Users/yn5/stuff/web/puha/keystone/admin/server/routes/list.js:13:12)
    at module.exports (/Users/yn5/stuff/web/puha/keystone/admin/server/routes/list.js:79:3)
    at Layer.handle [as handle_request] (/Users/yn5/stuff/web/puha/keystone/node_modules/express/lib/router/layer.js:95:5)
    at next (/Users/yn5/stuff/web/puha/keystone/node_modules/express/lib/router/route.js:131:13)
    at /Users/yn5/stuff/web/puha/keystone/admin/server/middleware/initList.js:9:7
    at Layer.handle [as handle_request] (/Users/yn5/stuff/web/puha/keystone/node_modules/express/lib/router/layer.js:95:5)
    at next (/Users/yn5/stuff/web/puha/keystone/node_modules/express/lib/router/route.js:131:13)
    at Route.dispatch (/Users/yn5/stuff/web/puha/keystone/node_modules/express/lib/router/route.js:112:3)
    at Layer.handle [as handle_request] (/Users/yn5/stuff/web/puha/keystone/node_modules/express/lib/router/layer.js:95:5)
    at /Users/yn5/stuff/web/puha/keystone/node_modules/express/lib/router/index.js:277:22
    at param (/Users/yn5/stuff/web/puha/keystone/node_modules/express/lib/router/index.js:349:14)
    at param (/Users/yn5/stuff/web/puha/keystone/node_modules/express/lib/router/index.js:365:14)
    at param (/Users/yn5/stuff/web/puha/keystone/node_modules/express/lib/router/index.js:365:14)
    at Function.process_params (/Users/yn5/stuff/web/puha/keystone/node_modules/express/lib/router/index.js:410:3)
    at next (/Users/yn5/stuff/web/puha/keystone/node_modules/express/lib/router/index.js:271:10)

**Comment by wuhaixing** [comment]
Me too

**Comment by wuhaixing** [comment]
The following code is from serialize-javascript/index.js:
  
      var IS_NATIVE_CODE_REGEXP = /\{\s*\[native code\]\s*\}/g;
      ......
      if (IS_NATIVE_CODE_REGEXP.test(serializedFn)) {
            throw new TypeError('Serializing native function: ' + fn.name);
       }

**Comment by snowkeeper** [comment]
Getting this error as well.

**Comment by ericelliott** [comment]
A minimal functional suite that at minimum could make requests on all our happy-path routes and verify there are no errors would be great. =)

**Comment by ericelliott** [comment]
A simple functional test to check a route for errors:

```js
import test from 'tape';
import request from 'supertest';
import keystone from 'test/helpers/keystone';

const app = keystone.app;
const route = '/';

test(`GET ${ route }`, assert => {
  assert.test(route, assert => {
    const msg = 'should not return an error';

    request(app)
      .get(route)
      .expect(200)
      .end(err => {
        assert.error(err, msg);
        assert.end();
      });
  });
});
```

**Comment by teaualune** [comment]
Apologize for not testing well, didn't notice that serialize-javascript treats native function serialization as an exception :sob:

@JedWatson do we need to prevent injection attacks like `<script>` tag injection while rendering values into jade? If all user inputs are already verified, then the library is not needed.

@ericelliott it looks cool! :+1:

**Comment by webteckie** [comment]
@teaualune jade is rendered server-side...is that still a problem with injection attacks?

**Comment by yn5** [comment]
Issue as reported is fixed by 82be01d125e06820c734622f307f8bfff6bfe33d, so I'm closing it. Maybe further discussion on the topic deserves a new issue. Thanks @JedWatson for the fix and all for chiming in!

**Comment by teaualune** [comment]
@webteckie the problem occurs when something unsafe being input to pre-rendered values. For example, if you define the `User` model with some plain string field (say, `someprop`) and then edit your user data to add something like `</script>XXX`, `JSON.stringify` with output something like this:

```
<script>
var Keystone = {};
Keystone.adminPath = '/keystone';
Keystone.backUrl = '/';
Keystone.user = {"someprop":"
</script>
XXX","email":...
```
Thus breaks the client-side codes.
